pages/blog/the-reference-rabbit-hole.md

```diff
@@ -0,0 +1,275 @@
+---
+title: "The Reference Rabbit Hole"
+date: 2021-11-18T12:00:00+01:00
+featured: true
+type: Engineering
+tags:
+  - Specification
+  - JSON Schema 
+cover: /img/posts/the-reference-rabbit-hole.webp
+authors:
+  - name: Jonas Lagoni
+    photo: /img/avatars/jonaslagoni.webp
+    link: https://github.com/jonaslagoni
+    byline: AsyncAPI Maintainer
+---
+
+So [Sergio](https://github.com/smoya) and I, went down a little bit of a rabbit hole the last couple of days when we were discussing [Fran's proposal to solve publish/subscribe confusion](https://github.com/asyncapi/spec/issues/618), and I thought I would share the journey.
+
+A lot of this can be seen as nitpicking... And I totally get this, as we need to venture deep into the specifications to fully understand the differences.
+
+I am gonna try to not use any complex words and explanations, so everyone can understand the problems, whether you are a novice AsyncAPI user or an experienced one.
+
+So let's split up the understanding of what references is, and where they can be used, and what is down in this rabbit hole. 
+
+# AsyncAPI references
+
+In AsyncAPI, we have something called a [Reference Object](https://www.asyncapi.com/docs/specifications/v2.2.0#referenceObject), which simply enables reusability in your AsyncAPI documents. This is possible through the simple keyword `$ref`. If we take a look at the [streetlight tutorial](https://www.asyncapi.com/docs/tutorials/streetlights), to utilize reusability, we could change [the document](https://www.asyncapi.com/docs/tutorials/streetlights#creating-the-asyncapi-file) to:
+
+```yaml
+asyncapi: '2.2.0'
+...
+channels:
+  light/measured:
+    publish:
+      summary: Inform about environmental lighting conditions for a particular streetlight.
+      operationId: onLightMeasured
+      message:
+        $ref: '#/components/messages/LightMeasured'
+components:
+  messages:
+    LightMeasured: 
+      name: LightMeasured
+      payload:
+        $ref: '#/components/schemas/LightMeasurement'
+  schemas:
+    LightMeasurement:
+      # Ignore the specifics here for now.
+```
+Here you can see that we simply reference where the definition of messages and payload schema is located.
+
+# Schema Object references
+As seen in the streetlight example, in AsyncAPI, to define your message payloads, we use use a [Schema Object](https://www.asyncapi.com/docs/specifications/v2.2.0#schemaObject), which is a superset of [JSON Schema draft 7](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01). 
+
+What `superset` means is we follow the JSON Schema draft 7 specification, but with a few modifications and additions to keywords.
```

Comment:
The reality is that we are wrongly saying we use a superset because for using a superset you first need to implement **all** elements of the set you extend.
In this case, we know there are some things we do not support.
I know the word is in current documentation, but I won't keep asserting Schema Object is a superset.
Maybe it's time to mention reality here?

Reply:
One thing is what tooling support, the other is what spec defines. The reality is that spec overwrites certain keywords, which then has a different meaning, like `$ref`. This technically renders `$id` unusable, but `$schema` should in theory still be supported as we define it as:

> The AsyncAPI Schema Object is a JSON Schema vocabulary which extends JSON Schema Core and Validation vocabularies. As such, any keyword available for those vocabularies is by definition available in AsyncAPI, and will work the exact same way, including but not limited to:

Reply:
My point is that I have the feeling we are wrongly saying it is a superset. By its definition:

> A set which includes another set or sets.
>
> Source: https://www.lexico.com/definition/superset

It does says anything about overriding or removing parts of the set, in fact I think this won't be ever expected. 😅